Shake/Main.lean

```diff
@@ -0,0 +1,243 @@
+import Lean.Util.FoldConsts
+import Lean
+
+open Lean
+
+abbrev Bitset := Nat
+
+structure State where
+  toIdx : HashMap Name USize := {}
+  modNames : Array Name := #[]
+  mods : Array ModuleData := #[]
+  deps : Array Bitset := #[]
+  transDeps : Array Bitset := #[]
+  needs : Array Bitset := #[]
+  constToIdx : HashMap Name USize := {}
+
+def calcNeeds (constToIdx : HashMap Name USize) (mod : ModuleData) : Bitset :=
+  mod.constants.foldl (init := 0) fun deps ci =>
+    let deps := visitExpr ci.type deps
+    match ci.value? with
+    | some e => visitExpr e deps
+    | none => deps
+where
+  visitExpr e arr :=
+    Lean.Expr.foldConsts e arr fun c arr => match constToIdx.find? c with
+      | some i => arr ||| (1 <<< i.toNat)
+      | none => arr
+
+unsafe def loadModules (imports : Array Import) : StateT State IO (Bitset × Bitset) := do
+  let mut imps := 0
+  let mut transImps := 0
+  for imp in imports do
+    let s ← get
+    if let some i := s.toIdx.find? imp.module then
+      imps := imps ||| (1 <<< i.toNat)
+      transImps := transImps ||| s.transDeps.uget i lcProof
+    else
+      let mFile ← findOLean imp.module
+      unless (← mFile.pathExists) do
+        throw <| IO.userError s!"object file '{mFile}' of module {imp.module} does not exist"
+      let (mod, _) ← readModuleData mFile
+      let (deps, transDeps) ← loadModules mod.imports
+      let s ← get
+      let n := s.mods.size.toUSize
+      let transDeps := transDeps ||| (1 <<< n.toNat)
+      imps := imps ||| (1 <<< n.toNat)
+      transImps := transImps ||| transDeps
+      set (σ := State) {
```

Comment:
Or perhaps
```suggestion
      setThe State {
```

Reply:
it doesn't exist

Reply:
Why do we have `getThe` but not `setThe`?

Reply:
probably not any good reason